### PR TITLE
Sticky Ad for NA mobile users

### PR DESCRIPTION
**Pull request by ripecosta**
## What does this change?
This PR adds a mobile sticky ad slot for the North American continent mobile users and puts it behind a 0% AB test.

Feature list:
- Mobile only;
- Articles only;
- North American only;
- Runs 320x50 ads;
- Ad refreshes every 30 seconds;
- Always visible at the bottom of the screen;
- Can be closed using the X button located in the Advertising label, but will come back every time the user refreshes the page;

Future developments:
- Enable prebid for this ad slot; (not in scope at this time)

## Screenshots
Mobile Sticky Ad 2019-05-29

## What is the value of this and can you measure success?
This will 

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [X] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [X] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Tested

- [X] Locally
- [X] On CODE (optional)

### CODE REVIEW

File: common/app/conf/switches/ABTestSwitches.scala
Comment (adamnfish) [inline common/app/conf/switches/ABTestSwitches.scala]: Is this description copied from an irrelevant test?

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (adamnfish) [inline static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: Is `closest` polyfilled for IE?

File: static/src/stylesheets/module/_adslot.scss
Comment (adamnfish) [inline static/src/stylesheets/module/_adslot.scss]: Should this classname be more BEM? Is this a `-container` or a `--container`? Same for the others here.

**Review by adamnfish — COMMENTED** [review]
Couple of little points but I'm no expert and it looks ok to me. Is this enough tests? Is it worth making sure the close button is included and working, for example?

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (adamnfish) [reply static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: Actually, it's fine in edge and I don't think we support IE anymore (?) so we're probably right as rain either way.

File: common/app/conf/switches/ABTestSwitches.scala
Comment (ripecosta) [reply common/app/conf/switches/ABTestSwitches.scala]: It is. Good catch. Thanks.

File: static/src/stylesheets/module/_adslot.scss
Comment (ripecosta) [reply static/src/stylesheets/module/_adslot.scss]: This particular one is totally BEM but the others need a second pass. Cheers.

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: Double checked both browsers both work as expected.

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (GHaberis) [inline static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: We actually have a number of standard Guardian `.svg`s that can be imported into your `.js` located here https://github.com/guardian/frontend/tree/master/static/src/inline-svgs/icon

Have you looked into using an Icon from here? I'd recommend this approach, they can be used in the following way:

`import crossIcon from 'svgs/icon/cross.svg';`

`closeDiv.innerHTML = `${crossIcon.markup}`;

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (GHaberis) [inline static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: This close `div` should really be a `button` as clicking it results in an actionable change to the page. A `button` element  gives an indication that this element is more than an element used for layout such as a `div`, this distinction is important as many accessibility tools parse the structure of a page to help guide their user through it and the elements used is important when denoting their  importance/role.

Here's some markup taken from another of our templates: 

```js
<button class="button engagement-banner__close-button js-site-message-close js-engagement-banner-close-button" data-link-name="hide release message">
  <span class="u-h">Close</span>
  ${closeCentralIcon.markup}
</button>
```
The `<span class="u-h">Close</span>` is  important as this  is picked up by screen readers, but isn't visible to users.

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (GHaberis) [inline static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: I'm surprised `flow` didn't pick up but it's possible `this.closest('.mobilesticky-container')` will return `null`, if this happens you'll get an error `Cannot read property 'remove' of null`, I'd therefore suggest  assigning `this.closest('.mobilesticky-container')` to a variable `container` (or something like that)  and checking it exists before removing it.

```js
const container = this.closest('.mobilesticky-container');

if (container) {
  container.remove();
}
```

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (GHaberis) [inline static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: It's not causing any issues but I think it's unnecessary to check `isInNA()` here as we're checking this in the `canShow` of the `commercialUsMobileSticky` test so there's no chance `isInVariantSynchronous(commercialUsMobileSticky, 'variant')` can be true if there not  in the US?

File: static/src/stylesheets/module/_adslot.scss
Comment (GHaberis) [inline static/src/stylesheets/module/_adslot.scss]: Can we use a colour code from the `_vars.scss` file? The colours in here should cover the entire palette used on the site.

Comment (ripecosta) [reply static/src/stylesheets/module/_adslot.scss]: 👍

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: 👍

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: We block screen readers on our ad elements, so apart from the close span 👍

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js]: 👍

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (katebee) [inline static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: 👋 Have you seen `createSlots` ?

```js
import { createSlots } from 'commercial/modules/dfp/create-slots';
```

**Comment by gtrufitt** [comment]
Coming from a platform perspective, it'd be good to know more of the success hypothesis of this ad. It's an invasive ad type and I'm concerned about the user experience.

Specifically - this ad is likely to have better impressions, likely to make more money but at what cost?

I'd prefer to also see a hypothesis on the negative impact of this ad type as it is the first type (I believe) that obscures content so we know what we're measuring before the A/B test goes out and how we know what success looks like. 

Thanks!

**Review by gtrufitt — CHANGES_REQUESTED** [review]
On another note, I can't see any checks to see if the acquisitions banner is showing? I don't think this is integrated with the [banner picker](https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/common/modules/ui/bannerPicker.js) so you're likely to get double banners, which is not great in itself.

**Comment by blongden73** [comment]
Can we not merge this please until someone has spoken to me about it...

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: 👍

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: This ad slot will only run on the North American continent. It's present here to make sure of that. And it's present again in AB test's canRun function to make sure it doesn't include users from outside the NA in the test.

**Comment by ripecosta** [comment]
We got the OK to go ahead and put this change behind a 0% AB test so it can be seen by stakeholders and tested from a UX point of view.

File: static/src/javascripts/projects/common/modules/experiments/tests/commercial-us-mobile-sticky.js
Comment (katebee) [inline static/src/javascripts/projects/common/modules/experiments/tests/commercial-us-mobile-sticky.js]: #### minor: 

`flow-strict` ? 

Using strict mode gives us some extra benefits: https://flow.org/en/docs/strict/

File: static/src/javascripts/projects/common/modules/experiments/tests/commercial-us-mobile-sticky.js
Comment (katebee) [inline static/src/javascripts/projects/common/modules/experiments/tests/commercial-us-mobile-sticky.js]: You may want to set the start this week if you plan to demo it or opt-in to the AB test and see this on PROD

Pretty sure that the canRun checks on AB tests adhere to the dates 😸

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (katebee) [inline static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: Do we have any tests for the mobile sticky? Even a couple would be helpful to document the intended behaviour

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (katebee) [inline static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: #### Minor / nitpick:

Is it mobilesticky, mobileSticky or mobile-sticky? 🙃 we are creating our slot with:

```js
const adSlot = createSlots('mobilesticky', {})[0];
```
Yet our slot is named `mobile-sticky`

**Review by katebee — COMMENTED** [review]
Added some comments but need to disappear for now 😄 
You may want to request another review from @gtrufitt

File: static/src/javascripts/projects/common/modules/experiments/tests/commercial-us-mobile-sticky.js
Comment (ripecosta) [reply static/src/javascripts/projects/common/modules/experiments/tests/commercial-us-mobile-sticky.js]: That was supposed to be 2019-05-23. Good catch. I'd be going mental trying to find what was wrong with it. Worked fine on CODE though.

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: I understand your nitpickyness and I agree. Changed. 👍

File: static/src/javascripts/projects/common/modules/experiments/tests/commercial-us-mobile-sticky.js
Comment (ripecosta) [reply static/src/javascripts/projects/common/modules/experiments/tests/commercial-us-mobile-sticky.js]: 👍

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (jeteve) [inline static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: Probably missing return. This function will always return the empty 'Promise.resolve()' from outside the if.

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (jeteve) [inline static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: Hu? Isn't that for the US only? Ha I think I get it. NA = North America, not 'Not Applicable'. Please comment so confusion is lifted.

**Review by jeteve — CHANGES_REQUESTED** [review]
See my 'missing return' comment.

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: Yes, NA stands for North America. Will leave a comment. 👍

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: 👍

**Review by jeteve — APPROVED** [review]

**Comment by jeteve** [comment]
@ripecosta happy for merging in the absence of outstanding code related requests.

File: static/src/javascripts/projects/commercial/modules/mobile-sticky.js
Comment (ripecosta) [reply static/src/javascripts/projects/commercial/modules/mobile-sticky.js]: We don't at the moment but will add them at a later date (team priorities). It will not be forgotten 🙂

**Pull request merged**
